snapcraft/plugins/_python/_pip.py

```diff
@@ -0,0 +1,263 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import collections
+import contextlib
+import json
+import logging
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+
+import snapcraft
+from snapcraft import file_utils
+from ._python_finder import (
+    get_python_command,
+    get_python_headers,
+)
+from . import errors
+
+logger = logging.getLogger(__name__)
+
+
+class Pip:
+    """Wrapper for pip abstracting the nasty args necessary for use in a part.
+
+    This class takes care of fetching pip, setuptools, and wheel, and then
+    simply shells out to pip with the magical arguments necessary to install
+    packages into a part. The important functions are:
+
+    download(): Fetch the provided list of packages (but do not install them)
+    install(): Install the provided list of packages (but do not fetch them)
+    wheel(): Build wheels for the provided list of packages (but do not fetch
+             them)
+    list(): List the installed packages
+    clean_packages(): Remove the package cache
+    env(): Obtain the environment required when calling pip
+
+    Of particular importance: packages must be downloaded (via download())
+    before they can be installed or have wheels built.
+    """
+
+    def __init__(self, *, python_major_version, part_dir, install_dir,
+                 stage_dir):
+        """Initialize pip.
+
+        First check to see if pip has already been installed. If not, fetch
+        pip, setuptools, and wheel, and install them so they can be used.
+
+        :param str python_major_version: The python major version to find (2 or
+                                         3)
+        :param str part_dir: Path to the part's working area
+        :param str install_dir: Path to the part's install area
+        :param str stage_dir: Path to the staging area
+
+        :raises MissingPythonCommandError: If no python could be found in the
+                                           staging or part's install area.
+        """
+        self._python_major_version = python_major_version
+        self._part_dir = part_dir
+        self._install_dir = install_dir
+        self._stage_dir = stage_dir
+
+        self._python_package_dir = os.path.join(self._part_dir, 'packages')
+        os.makedirs(self._python_package_dir, exist_ok=True)
+
+        self._python_command = get_python_command(
+            self._python_major_version, stage_dir=self._stage_dir,
+            install_dir=self._install_dir)
+
+        self._setup()
+
+    def _setup(self):
+        # Check to see if we have our own pip, yet. If not, we need to use the
+        # pip on the host (installed via build-packages) to grab our own.
+        try:
+            self._run([], stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as e:
```

Comment:
I know python is all about asking forgiveness and not permission, but this would look nicer to me with a function _is_pip_installed.

Reply:
indeed

Reply:
Done, thank you, that looks much better indeed.